import type { ControllerResponse } from "../../../../@types/responses";
import type { Tokens } from "../../../../@types/tokens";
import type { ExpressNextFunction, ExpressRequest } from "../../../../@types/wrappers";
import { PayloadHelper } from "../../../../app/helpers/PayloadHelper";
import type { IController } from "../../../../app/interfaces/IController";
import { HttpStatus, HttpStatusCode } from "../../../../app/schemas/HttpStatus";
import { ResponseUtil } from "../../../../app/utils/ResponseUtil";
import { AuthModule } from "../../../../modules/auth/module";
import { MyOrdersMessagesManager } from "./MyOrdersMessagesManager";
import { MyOrdersMessagesParams } from "./schemas/MyOrdersMessagesParams";
import { MyOrdersMessagesRequest } from "./schemas/MyOrdersMessagesRequest";
import type { MyOrdersMessagesResponse } from "./schemas/MyOrdersMessagesResponse";

export class MyOrdersMessagesController implements IController {
  public constructor(private readonly manager = new MyOrdersMessagesManager()) {}

  public async getMyOrdersMessages(
    req: ExpressRequest,
    res: ControllerResponse<MyOrdersMessagesResponse[] | null, Tokens | null>,
    next: ExpressNextFunction,
  ): Promise<typeof res | void> {
    try {
      // >-----------< AUTHORIZATION >-----------<
      const payload = PayloadHelper.getPayload(res);
      // >----------< VALIDATION >----------<
      const params = MyOrdersMessagesParams.parse(req);
      if (params.clientErrors.length > 0 || params.data === null) {
        return ResponseUtil.controllerResponse(
          res,
          new HttpStatus(HttpStatusCode.BAD_REQUEST),
          null,
          params.clientErrors,
          null,
          null,
        );
      }
      // >----------< LOGIC >----------<
      const out = await this.manager.getMyOrdersMessages(payload, params.data);
      // >----------< RESPONSE >----------<
      if (!out.httpStatus.isSuccess()) {
        return ResponseUtil.controllerResponse(
          res,
          out.httpStatus,
          out.serverError,
          out.clientErrors,
          out.data,
          null,
        );
      }
      return ResponseUtil.controllerResponse(
        res,
        out.httpStatus,
        out.serverError,
        out.clientErrors,
        out.data,
        await AuthModule.instance.refresh(payload),
      );
    } catch (error) {
      return next(error);
    }
  }

  public async postMyOrdersMessages(
    req: ExpressRequest,
    res: ControllerResponse<MyOrdersMessagesResponse[] | null, Tokens | null>,
    next: ExpressNextFunction,
  ): Promise<typeof res | void> {
    try {
      // >-----------< AUTHORIZATION >-----------<
      const payload = PayloadHelper.getPayload(res);
      // >----------< VALIDATION >----------<
      const params = MyOrdersMessagesParams.parse(req);
      if (params.clientErrors.length > 0 || params.data === null) {
        return ResponseUtil.controllerResponse(
          res,
          new HttpStatus(HttpStatusCode.BAD_REQUEST),
          null,
          params.clientErrors,
          null,
          null,
        );
      }
      const request = MyOrdersMessagesRequest.parse(req);
      if (request.clientErrors.length > 0 || request.data === null) {
        return ResponseUtil.controllerResponse(
          res,
          new HttpStatus(HttpStatusCode.BAD_REQUEST),
          null,
          request.clientErrors,
          null,
          null,
        );
      }
      // >----------< LOGIC >----------<
      const out = await this.manager.postMyOrdersMessages(payload, params.data, request.data);
      // >----------< RESPONSE >----------<
      if (!out.httpStatus.isSuccess()) {
        return ResponseUtil.controllerResponse(
          res,
          out.httpStatus,
          out.serverError,
          out.clientErrors,
          out.data,
          null,
        );
      }
      return ResponseUtil.controllerResponse(
        res,
        out.httpStatus,
        out.serverError,
        out.clientErrors,
        out.data,
        await AuthModule.instance.refresh(payload),
      );
    } catch (error) {
      return next(error);
    }
  }
}
